draft-ietf-tls-esni.md

```diff
@@ -435,25 +443,44 @@ The client-facing server computes ClientHelloInner by reversing this process.
 First it makes a copy of EncodedClientHelloInner and copies the
 legacy_session_id field from ClientHelloOuter. It then looks for an
 "ech_outer_extensions" extension. If found, it replaces the extension with the
-corresponding sequence of extensions in the ClientHelloOuter. If any referenced
-extensions are missing or if "encrypted_client_hello" appears in the list, the
-server MUST abort the connection with an "illegal_parameter" alert.
+corresponding sequence of extensions in the ClientHelloOuter. The server MUST
+abort the connection with an "illegal_parameter" alert if any of the following
+are true:
 
-The "ech_outer_extensions" extension is only used for compressing the
-ClientHelloInner. It MUST NOT be sent in either ClientHelloOuter or
-ClientHelloInner.
+* Any referenced extension is missing in ClientHelloOuter.
+
+* "encrypted_client_hello" appears in OuterExtensions.
+
+* OuterExtensions contains duplicate values.
 
-Note that it is possible to implement decoding of the EncodedClientHelloInner in
-a way that creates a denial-of-service vulnerability. Specifically, the server
-needs to check that each extension in the OuterExtensions list appears in the
-ClientHelloOuter. The naive strategy would require O(N\*M) time, where N is the
-number of extensions in the ClientHelloOuter and M is the number of extensions
-in the OuterExtensions list. Malicious clients could exploit this behavior in
-order to cause excessive work for the server, possibly making it unavailable.
-This problem can be mitigated by representing OuterExtensions in a way that
-allows it to be searched more quickly. For example, the runtime can be improved
-to O(N\*log(M)) by sorting the OuterExtensions and using binary search to access
-it.
+* The extensions in ClientHelloOuter corresponding to those in OuterExtensions
+  do not occur in the same order.
+
+If the cost to decode an EncodedClientHelloInner is disproportionately large in
+comparison to the input size, a malicious client could exploit this in a denial
+of service attack. Implementations are RECOMMENDED to use the following
+decoding procedure, which runs in linear time:
+
+1. Let `i` be zero and `n` be the number of extensions in ClientHelloOuter.
+
+1. For each extension type, `ext`, in OuterExtensions:
+
+   * If `ext` is "encrypted_client_hello", abort the connection with an
+     "illegal_parameter" alert and terminate this procedure.
+
+   * While the `i` is less than `n` and the `i`th extension of
+     ClientHelloOuter does not have type `ext`, increment `i`.
+
+   * If `i` is equal to `n`, abort the connection with an "illegal_parameter"
+     alert and terminate this procedure.
+
+   * Otherwise, the `i`th extension of ClientHelloOuter has type `ext`. Copy
```

Comment:
One nitpick: "`i`th" renders really badly in text.  Consider just avoiding the backticks.